### PR TITLE
V39

### PR DESCRIPTION
**Pull request by olegnn**

### CODE REVIEW

File: pallets/core_mods/src/migrations/mod.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        let mut did_controllers_removed = 0;
+
+        DidKeys::translate(|did, _: IncId, key: (PublicKey, u16)| {
+            did_keys_overridden += 1;
+            let key = UncheckedDidKey::new(key.0, VerRelType::try_from(key.1 & 0b1111).ok()?);
+
+            if let Ok(did_key) = key.clone().try_into() {
+                Some(did_key)
+            } else {
+                if key.ver_rels.intersects(VerRelType::CAPABILITY_INVOCATION) {
```
Comment (lovesh) [inline pallets/core_mods/src/migrations/mod.rs:29]: Why is the motivation behind this? Comment here would help.

File: pallets/core_mods/src/migrations/mod.rs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                        .and_then(identity)
+                    });
+                }
+
+                DidKey::new(key.public_key, VerRelType::KEY_AGREEMENT).ok()
+            }
+        });
+
+        let mut service_endpoints_overridden = 0;
+        DidServiceEndpoints::translate_values(|(types, origins): (u16, Vec<WrappedBytes>)| {
```
Comment (lovesh) [inline pallets/core_mods/src/migrations/mod.rs:62]: Why is this needed?

File: pallets/core_mods/src/modules/did/base/mod.rs [hunk 3]
```diff
@@ -91,3 +94,27 @@ impl<T: Config + Debug> Module<T> {
         Dids::<T>::insert(did, did_details.into())
     }
 }
+
+impl<T: Config> scale_info::TypeInfo for StoredDidDetails<T> {
```
Comment (lovesh) [inline pallets/core_mods/src/modules/did/base/mod.rs:98]: Why would derive `scale_info::TypeInfo` won't work?  Because of `VerRelType`?

File: pallets/core_mods/src/modules/did/keys.rs [hunk 4]
```diff
@@ -192,43 +297,85 @@ impl<T: Config + Debug> Module<T> {
...
-
-                key
-            };
-            if key.can_control() {
-                controller_keys_count += 1;
-            }
-
-            keys_to_insert.push(key);
-        }
+impl scale_info::TypeInfo for DidKey {
```
Comment (lovesh) [inline pallets/core_mods/src/modules/did/keys.rs:302]: Same question as before, why derive won't work?

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -612,26 +644,125 @@ parameter_types! {
         .saturating_sub(BlockExecutionWeight::get());
 }
 
+pub struct BaseFeeThreshold;
```
Comment (lovesh) [inline runtime/src/lib.rs:647]: Some comments explaining these would be helpful, like what are lower, ideal, elasticity, etc and why these values?/

File: node/src/chain_spec.rs [hunk 6]
```diff
@@ -391,8 +410,165 @@ pub fn pos_testnet_config() -> ChainSpec {
...
+                    ),
+                ],
+                endowed_accounts: [
+                    "3At9HRmfpkJmkpzznNoy3o1zC7BUfskgfkKTRqPC2W8vnmaD",
+                    "37AcAUVB2aai9MMQrdeiPV5MQbDpZMKsFDqdMkfkA1xPUzLw",
+                    "3BUtge3oi9PUPT2F8C419sTfK5ApiyiFCiz6JivoZNmnmJ9E",
+                    "38ty5VuuQzzyRVYbqZSWiK6ek531H8qJJiF5RPnEgmHZE5z1",
+                    "3BWu6h4Cge1AQTjqvBHnbqBhMShUtXhZgHaDkYs43hMhdgyR",
+                    "3AFuSQTueBLr7y7yg5jNXEaWUJNYDiTVf16QXznA3r78UM7N",
+                    "391LjoiCr1JjsrqHJW6EirNFUSjixdNGfkVDi68Uibgax8Zt", // Beefy system treasury
```
[the remark this answers is not in the document]
Comment (olegnn) [reply node/src/chain_spec.rs:496]: For system `beefy` payouts. It's the exact `devnet` configuration as used by the bridge, and we can omit it for now.

File: node/src/chain_spec.rs [hunk 7]
```diff
@@ -661,7 +839,7 @@ impl GenesisBuilder {
                     })
                     .collect::<Vec<_>>(),
             },
-            poa: PoAModuleConfig {
+            po_a_module: PoAModuleConfig {
```
[the remark this answers is not in the document]
Comment (olegnn) [reply node/src/chain_spec.rs:842]: I guess some internal macro code in `substrate` was modified, and that resulted in naming changes.

File: pallets/bridge/merkle-proof/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+[package]
```
[the remark this answers is not in the document]
Comment (olegnn) [reply pallets/bridge/merkle-proof/Cargo.toml:1]: We don't have this module in any public repo yet (except for the original `snowbridge`). 
I can either reference their repo with a specific commit or publish this pallet in a separate repo, but I guess this will only complicate things.

File: pallets/core_mods/src/modules/did/base/onchain.rs [hunk 9]
```diff
@@ -94,9 +100,14 @@ impl<T: Config + Debug> Module<T> {
     ) -> Result<(), Error<T>> {
         // This will result in the removal of DID from storage map `Dids`
         details.take();
-        DidKeys::remove_prefix(did);
-        DidControllers::remove_prefix(did);
-        DidServiceEndpoints::remove_prefix(did);
+        // TODO: limit and cursor
```
[the remark this answers is not in the document]
Comment (olegnn) [reply pallets/core_mods/src/modules/did/base/onchain.rs:103]: The idea is to limit the number of entities to be removed in a single call. If some entities are remaining, the caller should call removal again. This way we can fully control transaction weight.
I tried to make it work this way before but in the old `substrate` `remove_prefix` method didn't even give any information about how many entities were removed, not to say about cursors.

File: pallets/core_mods/src/modules/did/base/mod.rs (hunk 3 above)
Comment (olegnn) [reply pallets/core_mods/src/modules/did/base/mod.rs:98]: Because there's no other way to omit the type name prefix constructing from the file path.

File: pallets/core_mods/src/migrations/mod.rs (hunk 1 above)
Comment (olegnn) [reply pallets/core_mods/src/migrations/mod.rs:29]: Added a comment

File: runtime/src/lib.rs (hunk 5 above)
Comment (olegnn) [reply runtime/src/lib.rs:654]: Added docs

File: pallets/bridge/merkle-proof/Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/bridge/merkle-proof/Cargo.toml:1]: I was talking about the public snowbridge repo. Whats the complexity in referencing it with a commit?

File: pallets/core_mods/src/modules/did/base/onchain.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (lovesh) [reply pallets/core_mods/src/modules/did/base/onchain.rs:103]: Using multiple calls will lead to a situation where DID will be removed but some keys will still be pending unless further calls are made. Or we don't remove DID details unless all other data is deleted. Maybe introduce an upper bound on the number of keys, endspoints, etc can be created and on removal require fees to cover the max and then return extra fee at the end.

File: pallets/core_mods/src/migrations/mod.rs (hunk 1 above)
Comment (lovesh) [reply pallets/core_mods/src/migrations/mod.rs:31]: I don't follow, can you please add a more detailed comment?

File: pallets/core_mods/src/migrations/mod.rs (hunk 2 above)
Comment (lovesh) [reply pallets/core_mods/src/migrations/mod.rs:65]: Can you please add a more detailed comment here as well?

File: pallets/core_mods/src/modules/did/base/mod.rs (hunk 3 above)
Comment (lovesh) [reply pallets/core_mods/src/modules/did/base/mod.rs:98]: Which "type name prefix"? Where do we construct it from file path?

File: runtime/src/lib.rs (hunk 5 above)
Comment (lovesh) [reply runtime/src/lib.rs:654]: Thanks

File: pallets/core_mods/src/modules/did/base/mod.rs (hunk 3 above)
Comment (olegnn) [reply pallets/core_mods/src/modules/did/base/mod.rs:98]: Metadata V14 type name - it will be prefixed with the module path, see https://github.com/paritytech/scale-info/blob/master/derive/src/lib.rs#L124. 
We discussed it before but always can rename `Did` to `CoreModsModulesDidBaseDid` in `SDK`. 
As I already mentioned in our chat, a more appropriate solution would be to reorganize modules and define all prelude types in some root crate and build a correct tree structure. For ex., rename types like `Did` -> `DockDid`, `Controller` -> `DockDidController` etc.

File: pallets/bridge/merkle-proof/Cargo.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (olegnn) [reply pallets/bridge/merkle-proof/Cargo.toml:1]: I guess it's better to minimize dependencies to some uncontrolled repo, as it can be moved, history may be overridden, and the repository can be hidden - any such move will break the build.
But anyway I will move this pallet to dependencies to reduce the amount of duplicate code.

**Review by lovesh — APPROVED** [review]
